### PR TITLE
Feature/teleoperation tests

### LINKED ISSUE #27: [6D teleop] test case
simple test case:

- [ ] driver outputs twist message
- [ ] ...

### PR DESCRIPTION
**Pull request by DamKoVosh**
Testsetup for teleoperation package.

Closes #27.

### CODE REVIEW

**Comment by DamKoVosh** [comment]
To manually run tests:
```
PYTHONPATH=$PYTHONPATH:"/usr/lib/python2.7/dist-packages"
catkin_make -DENABLE_COVERAGE_TESTING=ON -DCMAKE_BUILD_TYPE=Debug
catkin_make -DENABLE_COVERAGE_TESTING=ON -DCMAKE_BUILD_TYPE=Debug pilz_teleoperation_coverage
```

**Comment by DamKoVosh** [comment]
At least manual code coverage analysis still has a problem.
Getting 
```
Processing code coverage counters and generating report. Find the report in /home/ri/.ros/htmlcov/index.html
Name    Stmts   Miss  Cover
---------------------------
No data to report.
pilz_teach/pilz_teleoperation/CMakeFiles/pilz_teleoperation_coverage.dir/build.make:57: recipe for target 'pilz_teach/pilz_teleoperation/CMakeFiles/pilz_teleoperation_coverage' failed
make[3]: *** [pilz_teach/pilz_teleoperation/CMakeFiles/pilz_teleoperation_coverage] Error 1
CMakeFiles/Makefile2:1889: recipe for target 'pilz_teach/pilz_teleoperation/CMakeFiles/pilz_teleoperation_coverage.dir/all' failed
make[2]: *** [pilz_teach/pilz_teleoperation/CMakeFiles/pilz_teleoperation_coverage.dir/all] Error 2
CMakeFiles/Makefile2:1896: recipe for target 'pilz_teach/pilz_teleoperation/CMakeFiles/pilz_teleoperation_coverage.dir/rule' failed
make[1]: *** [pilz_teach/pilz_teleoperation/CMakeFiles/pilz_teleoperation_coverage.dir/rule] Error 2
Makefile:825: recipe for target 'pilz_teleoperation_coverage' failed
make: *** [pilz_teleoperation_coverage] Error 2

```

**Comment by DamKoVosh** [comment]
The only py analysed at the moment is tst_key_teleop.py. It does not recognize that this test starts the driver and the nosetests are ignored at all.

**Comment by jschleicher** [comment]
the `_nosetests` macro reads the environment variable `CATKIN_TEST_COVERAGE`, so that has to be set, see [catkin/cmake](https://github.com/ros/catkin/blob/61c403ffab2c4ca197f27dc4c90f791b049cc8a9/cmake/test/nosetests.cmake#L59-L60)

```
export CATKIN_TEST_COVERAGE=1
```
Then a `.coverage` is created inside the build directory. We'd have to combine them with the rostest output and different nosetest runs (that are using unittest without nose by the way)

In fact I think we probably need to fix the catkin cmake macros for proper python coverage...

**Comment by jschleicher** [comment]
I'd suggest to merge this branch without coverage checks for now. This still adds lint PEP8-check + basic test(s). Coverage moved into #39.

File: pilz_teleoperation/test/integrationtests/tst_key_teleop.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    def tearDown(self):
+        rospy.loginfo("TearDown called...")
+
+    def callback(self, msg):
+        self.twist_published = True
+
+    def test_startup(self):
+        """ Test if driver gets started """
+        subscriber_twist = rospy.Subscriber('/jog_server/delta_jog_cmds', TwistStamped, self.callback)
+        subscriber_joint = rospy.Subscriber('/jog_server/joint_delta_jog_cmds', JointJog, self.callback)
```
Comment (jschleicher) [inline pilz_teleoperation/test/integrationtests/tst_key_teleop.py:44]: This simply tests publication, right? I'm not sure if that is worth a self-written test file; could be accomplished using [hztest](http://wiki.ros.org/rostest/Nodes) as well... But probably this is just a starting point and will add more checks in the future??

File: prbt_jog_arm_support/test/integrationtests/tst_jog_arm_motion.py [hunk 2]
```diff
@@ -65,4 +65,4 @@ def test_move(self):
 if __name__ == '__main__':
     import rostest
     rospy.init_node('tst_jog_arm_motion')
-    rostest.rosrun('pilz_jog_arm_suppport', 'tst_jog_arm_motion', TestJogArmMotion)
+    rostest.rosrun('prbt_jog_arm_suppport', 'tst_jog_arm_motion', TestJogArmMotion)
```
Comment (Hardthof) [inline prbt_jog_arm_support/test/integrationtests/tst_jog_arm_motion.py]: prbt_jog_arm_suppport has cute triplets (ppp)

File: pilz_teleoperation/test/integrationtests/tst_key_teleop.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+        timeout = rospy.Time().now() + rospy.Duration(secs=10)
+        while not rospy.is_shutdown() and not self.twist_published:
+            rospy.sleep(0.1)
+            self.assertTrue(rospy.Time().now() < timeout, "driver did not publish")
+        subscriber_twist.unregister()
+        subscriber_joint.unregister()
```
Comment (Hardthof) [inline pilz_teleoperation/test/integrationtests/tst_key_teleop.py:51]: why unregister?

File: pilz_teleoperation/test/integrationtests/tst_key_teleop.py (hunk 1 above)
Comment (DamKoVosh) [reply pilz_teleoperation/test/integrationtests/tst_key_teleop.py:44]: yes, it was just a simple way to test the test structure and do anything useful in the meantime. This should be improved in the future. I would however suggest we don't do it in this pull request, to keep the merge small and have the test structure in the main branch already

File: pilz_teleoperation/test/integrationtests/tst_key_teleop.py (hunk 3 above)
Comment (DamKoVosh) [reply pilz_teleoperation/test/integrationtests/tst_key_teleop.py:51]: Not 100% sure if its necessary here, but when i did tests in earlier projects i ran into the problem that when running multiple tests, the nodes did not shut down properly/fast enough for the next test. Unregistering ensures we don't throw warnings in that case.

Comment (DamKoVosh) [reply pilz_teleoperation/test/integrationtests/tst_key_teleop.py:51]: If that is not or no longer the case tell me, then we can remove them.

**Review by tobi-v — APPROVED** [review]

**Pull request merged**